README.adoc

```diff
@@ -0,0 +1,1763 @@
+&bull; `context` blocks should pretty much always have an opposite negative case.
+It is a code smell if there is a single context (without a matching
+negative case), and this code needs refactoring, or may have no purpose.
+^[<<context-cases,link>>]^
+
+[source,ruby]
+----
+# bad - needs refactoring
+describe '#attributes' do
+  context 'the returned hash' do
+    it 'includes the display name' do
+      # ...
+    end
+
+    it 'includes the creation time' do
+      # ...
+    end
+  end
+end
+
+# bad - the negative case needs to be tested, but isn't
+describe '#attributes' do
+  context 'when display name is present' do
+    before do
+      subject.display_name = 'something'
+    end
+
+    it 'includes the display name' do
+      # ...
+    end
+  end
+end
+
+# good
+describe '#attributes' do
+  subject { FactoryBot.create(:article) }
+
+  specify do
+    expect(subject.attributes).to include subject.display_name
+    expect(subject.attributes).to include subject.created_at
+  end
+end
+
+describe '#attributes' do
+  context 'when display name is present' do
+    before do
+      subject.display_name = 'something'
+    end
+
+    it 'includes the display name' do
+      # ...
+    end
+  end
+
+  context 'when display name is not present' do
+    before do
+      subject.display_name = nil
+    end
+
+    it 'does not include the display name' do
+      # ...
+    end
+  end
+end
+----
+
+'''
+
+[[let-blocks]]
+&bull; Use `let` and `let!` for data that is used across several examples in an
+example group. Use `let!` to define variables even if they are not
+referenced in some of the examples, e.g. when testing balancing negative
+cases. Do not overuse ``let``s for primitive data, find the balance between
+frequency of use and complexity of the definition.
+^[<<let-blocks,link>>]^
+
+[source,ruby]
+----
+# bad
+it 'finds shortest path' do
+  tree = Tree.new(1 => 2, 2 => 3, 2 => 6, 3 => 4, 4 => 5, 5 => 6)
+  expect(dijkstra.shortest_path(tree, from: 1, to: 6)).to eq([1, 2, 6])
+end
+
+it 'finds longest path' do
+  tree = Tree.new(1 => 2, 2 => 3, 2 => 6, 3 => 4, 4 => 5, 5 => 6)
+  expect(dijkstra.longest_path(tree, from: 1, to: 6)).to eq([1, 2, 3, 4, 5, 6])
+end
+
+# good
+let(:tree) { Tree.new(1 => 2, 2 => 3, 2 => 6, 3 => 4, 4 => 5, 5 => 6) }
+
+it 'finds shortest path' do
+  expect(dijkstra.shortest_path(tree, from: 1, to: 6)).to eq([1, 2, 6])
+end
+
+it 'finds longest path' do
+  expect(dijkstra.longest_path(tree, from: 1, to: 6)).to eq([1, 2, 3, 4, 5, 6])
+end
+----
+
+* +++<a name="instance-variables">++++++</a>+++
+Use `let` definitions instead of instance variables.
+^[<<instance-variables,link>>]^
++
+[source,ruby]
+----
+# bad
+before { @name = 'John Wayne' }
+
+it 'reverses a name' do
+  expect(reverser.reverse(@name)).to eq('enyaW nhoJ')
+end
+
+# good
+let(:name) { 'John Wayne' }
+
+it 'reverses a name' do
+  expect(reverser.reverse(name)).to eq('enyaW nhoJ')
+end
+----
+
+* +++<a name="shared-examples">++++++</a>+++
```

Comment:
There's no need to use raw HTML to make an anchor on a list. You can write it as follows:

```
* [[shared-examples]]Use shared examples to reduce duplication.
```